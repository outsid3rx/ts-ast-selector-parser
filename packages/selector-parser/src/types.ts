export interface AttributeCondition {
  type: 'attribute'
  attribute: string
  operator: string
  value: string
}

export interface TagCondition {
  type: 'tag'
  name: string
}

export interface Step {
  node: TagCondition
  attributes: AttributeCondition[]
}

export type Combinator = 'descendant' | 'child'

export interface SelectorPlan {
  steps: Step[]
  combinators: Combinator[]
}
