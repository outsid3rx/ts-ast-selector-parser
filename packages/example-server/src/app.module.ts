import { Module } from '@nestjs/common'
import { ServeStaticModule } from '@nestjs/serve-static'
import { join } from 'node:path'

import { MatcherModule } from './matcher/matcher.module'

@Module({
  imports: [
    MatcherModule,
    ServeStaticModule.forRoot({
      rootPath: join(__dirname, '../..', 'example/dist'),
    }),
  ],
})
export class AppModule {}
