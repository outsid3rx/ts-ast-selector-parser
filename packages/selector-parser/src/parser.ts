import postcssSelectorParser from 'postcss-selector-parser'

import { Combinator, SelectorPlan, Step } from './types'

export function parseSelectorToPlan(selectorStr: string): SelectorPlan {
  const root = postcssSelectorParser().astSync(selectorStr)

  const [selector] = root.nodes
  if (!selector || selector.type !== 'selector') {
    throw new Error('Invalid or empty selector')
  }

  const steps: Step[] = []
  const combinators: Combinator[] = []

  let currentStep: Step = { node: null, attributes: [] }

  for (const node of selector.nodes) {
    if (node.type === 'combinator') {
      if (!currentStep.node) {
        throw new Error('Combinator must follow a tag')
      }
      steps.push(currentStep)

      const value = node.value.trim()
      if (value === '>') {
        combinators.push('child')
      } else if (value === '') {
        combinators.push('descendant')
      } else {
        throw new Error(`Unsupported combinator: "${value}"`)
      }

      currentStep = { node: null, attributes: [] }
    } else if (node.type === 'tag') {
      currentStep.node = { type: 'tag', name: node.value }
    } else if (node.type === 'attribute') {
      if (node.operator !== '=') {
        throw new Error(
          `Only "=" attribute operator is supported, got: ${node.operator}`,
        )
      }
      if (typeof node.value !== 'string') {
        throw new TypeError('Only string attribute values are supported')
      }
      currentStep.attributes.push({
        type: 'attribute',
        attribute: node.attribute,
        operator: node.operator,
        value: node.value,
      })
    }
  }

  if (currentStep.node) {
    steps.push(currentStep)
  } else {
    throw new Error('Selector ends with combinator or is malformed')
  }

  return { steps, combinators }
}
