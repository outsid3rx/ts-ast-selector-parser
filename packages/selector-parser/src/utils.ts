import ts from 'typescript'

import { Step } from './types'

export const getChildren = (node: ts.Node) => {
  const children: ts.Node[] = []
  ts.forEachChild(node, (child) => {
    children.push(child)
  })
  return children
}

export const stringToSyntaxKind = (kindName: string): ts.SyntaxKind => {
  const kind = ts.SyntaxKind[kindName]
  if (kind === undefined) {
    throw new Error(`Unknown SyntaxKind: ${kindName}`)
  }
  return kind
}

export const getNodeProperty = (node: ts.Node, prop: string): string => {
  // Не для всех свойств есть прямое соответствие
  if (prop === 'name') {
    if ('name' in node) {
      const nameNode = node.name as ts.Identifier
      if (ts.isIdentifier(nameNode)) {
        return nameNode.text
      }
    }
    return undefined
  }

  // По умолчанию — просто читаем свойство
  return node[prop]
}

export const matchesStep = (step: Step, node: ts.Node) => {
  const expectedKind = stringToSyntaxKind(step.node.name)
  if (node.kind !== expectedKind) return false

  for (const attr of step.attributes) {
    const actualValue = getNodeProperty(node, attr.attribute)
    if (actualValue !== attr.value) {
      return false
    }
  }

  return true
}
