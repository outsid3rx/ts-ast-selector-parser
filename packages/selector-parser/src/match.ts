import ts from 'typescript'

import { SelectorPlan } from './types'
import { getChildren, matchesStep } from './utils'

export const querySelectorAll = (plan: SelectorPlan, root: ts.Node) => {
  const results: ts.Node[] = []

  function matchFrom(node: ts.Node, stepIndex: number): boolean {
    if (stepIndex >= plan.steps.length) return false

    if (!matchesStep(plan.steps[stepIndex], node)) {
      return false
    }

    if (stepIndex === plan.steps.length - 1) {
      results.push(node)
      return true
    }

    const nextCombinator = plan.combinators[stepIndex]
    const directChildren = getChildren(node)

    if (nextCombinator === 'child') {
      for (const child of directChildren) {
        matchFrom(child, stepIndex + 1)
      }
    } else if (nextCombinator === 'descendant') {
      // Обходим всё поддерево
      walkDescendant(node, stepIndex)
    }

    return true
  }

  const walkDescendant = (n: ts.Node, stepIndex: number) => {
    const kids = getChildren(n)
    for (const k of kids) {
      matchFrom(k, stepIndex + 1)
      walkDescendant(k, stepIndex)
    }
  }

  const walk = (n: ts.Node) => {
    matchFrom(n, 0)
    const children = getChildren(n)
    for (const child of children) {
      walk(child)
    }
  }

  walk(root)
  return results
}
