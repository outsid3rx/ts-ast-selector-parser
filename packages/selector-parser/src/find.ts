import ts from 'typescript'

import { querySelectorAll } from './match'
import { parseSelectorToPlan } from './parser'

export const parseAndFind = (selector: string, source: string) => {
  const sourceFile = ts.createSourceFile(
    'temp.ts',
    source,
    ts.ScriptTarget.Latest,
    true,
  )

  const plan = parseSelectorToPlan(selector)
  return querySelectorAll(plan, sourceFile)
}
