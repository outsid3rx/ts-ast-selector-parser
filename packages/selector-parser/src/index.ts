import { parseAndFind } from './find'
import { querySelectorAll } from './match'
import { parseSelectorToPlan } from './parser'

export { parseAndFind, parseSelectorToPlan, querySelectorAll }
