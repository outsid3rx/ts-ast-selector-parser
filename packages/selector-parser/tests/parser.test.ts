import { expect, test } from 'vitest'

import { parseSelectorToPlan } from '../src'
import { MOCK_SELECTOR } from './mocks'

test('Parser should create plan', () => {
  expect(parseSelectorToPlan(MOCK_SELECTOR)).toMatchSnapshot()
})
