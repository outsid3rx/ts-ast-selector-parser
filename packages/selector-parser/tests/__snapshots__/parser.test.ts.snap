// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Parser should create plan 1`] = `
{
  "combinators": [
    "child",
  ],
  "steps": [
    {
      "attributes": [],
      "node": {
        "name": "ClassDeclaration",
        "type": "tag",
      },
    },
    {
      "attributes": [
        {
          "attribute": "name",
          "operator": "=",
          "type": "attribute",
          "value": "init",
        },
      ],
      "node": {
        "name": "MethodDeclaration",
        "type": "tag",
      },
    },
  ],
}
`;
